_posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md

```diff
@@ -0,0 +1,98 @@
+A header is basically just a key and a value, and in case of JWT, the value is encoded.
+Therefore it's hard to do versioning or any kind of content management on the data transferred inside these tokens.
+
+In the aforementioned example, the JWT token couples the products microservice with the orders microservice.
+If the products microservice changes the structure of the JWT token, the orders microservice will no longer be able to read it.
+While this coupling would exist as well when the orders microservice would call the products microservice directly, we would manage that coupling as part of the contract between these two microservices.
+In our case we don't know there is a link between the two microservices since they don't call each other directly.
+Yet by transferring the JWT token from one microservice over the client to the other microservice, we are creating a hidden dependency.
+It's also hard to have versioning on headers unless we put the version inside the name of the header.
+
+### Scaling
+
+Adding versions to the headernames, documenting which microservices expect which versions of JWT tokens of other microservices, and making sure we implement the tolerant-reader principle when reading the tokens might be a step in the right direction to avoid mass hysteria when tokens have to be adjusted.
+But what is simply impossible to get around is the size restriction of headers in HTTP requests and responses.
+The HTTP specification doesn't put any restriction on header size (singular or combined).
+But webservers, reverse proxies, CDNs and other network components do.
+Why they do this is not entirely clear as the spec allows any size, but the fact of the matter is that the restrictions exist.
+Putting a list of ids in a header like in our products example, will eventually break as the list could get too long.
+It's not even clear how long is too long.
+
+## Alternatives
+
+We see three possible alternatives to this failed approach to manage state.
+
+Instead of passing the state from one microservice over a client to another microservice, we could pass the state as part of the body of the request and response.
+The downside of this approach is that we can no longer use GET methods for the calls where we need to pass the previously fetched state.
+
+The second alternative is to persist the state in a key value datastore on the server.
+We could asynchronously fetch products data and store it inside a datastore owned by the orders microservice.
+This could get stale, but so could a cache on the products microservice.
+This approach seems most common in the industry and could be well be the most preferrable.
+
+And when all else fails, we can still simply make a call from the orders microservice to the products microservice and count on caching.
+
+## Conclusion
+
+Using JWT tokens as a means to transfer state to and from microservices via the client could seem like a good idea, but in the end turned out to be quite an anti-pattern.
```

Comment:
Using JWT tokens as a means to transfer state to and from microservices via the client seemed like a good idea